ksqldb-engine/src/main/java/io/confluent/ksql/planner/plan/SelectionUtil.java

```diff
@@ -51,24 +87,46 @@ public static LogicalSchema buildProjectionSchema(
         functionRegistry
     );
 
-    final Builder builder = LogicalSchema.builder();
-
-    final ImmutableMap.Builder<ColumnName, SqlType> keys = ImmutableMap.builder();
+    // keyExpressions[i] represents the expressions found in projection
+    // that are associated with parentSchema's key at index i
+    final List<List<SelectExpression>> keyExpressions = new ArrayList<>(parentSchema.key().size());
+    for (int i = 0; i < parentSchema.key().size(); i++) {
+      keyExpressions.add(new ArrayList<>());
+    }
 
+    // first pass to construct keyExpressions, keyExpressionMembership
+    // is just a convenience data structure so that we don't have to do
+    // the isKey check in the second iteration below
+    final Set<SelectExpression> keyExpressionMembership = new HashSet<>();
     for (final SelectExpression select : projection) {
       final Expression expression = select.getExpression();
+      if (expression instanceof ColumnReferenceExp) {
+        final ColumnName name = ((ColumnReferenceExp) expression).getColumnName();
+        parentSchema.findColumn(name)
+            .filter(c -> c.namespace() == Namespace.KEY)
+            .ifPresent(c -> {
+              keyExpressions.get(c.index()).add(select);
+              keyExpressionMembership.add(select);
+            });
+      }
+    }
 
-      final SqlType expressionType = expressionTypeManager
-          .getExpressionSqlType(expression);
-
-      final boolean keyColumn = expression instanceof ColumnReferenceExp
-          && parentSchema.isKeyColumn(((ColumnReferenceExp) expression).getColumnName());
-
-      if (keyColumn) {
-        builder.keyColumn(select.getAlias(), expressionType);
-        keys.put(select.getAlias(), expressionType);
+    // second pass, which iterates the projections but ignores any key expressions,
+    // instead taking them from the ordered keyExpressions list
+    final Builder builder = LogicalSchema.builder();
+    int currKeyIdx = 0;
+    for (final SelectExpression select : projection) {
+      if (keyExpressionMembership.contains(select)) {
+        final SelectExpression keyExp = keyExpressions.get(currKeyIdx).remove(0);
+        final SqlType type = expressionTypeManager.getExpressionSqlType(keyExp.getExpression());
+        builder.keyColumn(keyExp.getAlias(), type);
+        if (keyExpressions.get(currKeyIdx).isEmpty()) {
```

Comment:
What happens if a key isn't selected? (Should this be a while-loop rather than a simple if-check?  If so, we also need to update `currKeyIdx = 0;` on line 117.)

Reply:
good call - made this a while loop and moved it up (before the remove) so no need to change line 117